src/GafferScene/BranchCreator.cpp

```diff
@@ -80,6 +82,99 @@ void mergeSetNames( const InternedStringVectorData *toAdd, vector<InternedString
 
 } // namespace
 
+//////////////////////////////////////////////////////////////////////////
+// BranchCreator::BranchesData
+//////////////////////////////////////////////////////////////////////////
+
+class BranchCreator::BranchesData : public IECore::Data
+{
+
+	public :
+
+		BranchesData( const BranchCreator *branchCreator, const Context *context )
+		{
+			auto f = [this, branchCreator]( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+			{
+				addBranch( branchCreator, path );
+				return true;
+			};
+			SceneAlgo::filteredParallelTraverse( branchCreator->inPlug(), branchCreator->filterPlug(), f );
+
+			if( !branchCreator->filterPlug()->getInput() )
+			{
+				const auto parent = branchCreator->parentPlugPath();
+				if( parent )
+				{
+					addBranch( branchCreator, *parent );
+				}
+			}
+		}
+
+		static bool affectedBy( const BranchCreator *branchCreator, const Plug *input )
+		{
+			return
+				input == branchCreator->filterPlug() ||
+				input == branchCreator->inPlug()->childNamesPlug() ||
+				input == branchCreator->parentPlug() ||
+				input == branchCreator->inPlug()->existsPlug()
+			;
+		}
+
+		static void hash( const BranchCreator *branchCreator, const Context *context, IECore::MurmurHash &h )
+		{
+			// See `SceneAlgo::matchingPathsHash()` for documentation of this hashing strategy.
+			std::atomic<uint64_t> h1, h2;
+			auto f = [branchCreator, &h1, &h2]( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+			{
+				IECore::MurmurHash h;
+				hashBranch( branchCreator, path, h );
+				h1 += h.h1();
+				h2 += h.h2();
+				return true;
+			};
+			SceneAlgo::filteredParallelTraverse( branchCreator->inPlug(), branchCreator->filterPlug(), f );
+			h.append( MurmurHash( h1, h2 ) );
+
+			if( !branchCreator->filterPlug()->getInput() )
+			{
+				const auto parent = branchCreator->parentPlugPath();
+				if( parent )
+				{
+					hashBranch( branchCreator, *parent, h );
+				}
+			}
+		}
+
+		const IECore::PathMatcher &parentPaths() const
+		{
+			return m_parentPaths;
+		}
+
+	private :
+
+		static void hashBranch( const BranchCreator *branchCreator, const ScenePlug::ScenePath &path, IECore::MurmurHash &h )
+		{
+			if( path.size() )
+			{
+				h.append( path.data(), path.size() );
+			}
+			else
+			{
+				h.append( 0 );
+			}
+		}
+
+		void addBranch( const BranchCreator *branchCreator, const ScenePlug::ScenePath &path )
+		{
+			tbb::spin_mutex::scoped_lock lock( m_mutex );
```

Comment:
Oh, this optimization is no longer an option once we add support for destinationPlug.  Nevermind.

Reply:
Yep, wouldn't work here for that reason. I do have a (git) branch where I do the lock-free building of a PathMatcher in `SceneAlgo::matchingPaths()` though, and will get it polished up for review at some point. My performance tests for it were all based round a huge instancer though, and I think we need that other parallel-for-in-parallel-traverse optimisation we discussed before the lock becomes the primary issue.

It would probably be possible to do a read-write lock per location in this code, but since this branch is primarily about adding a feature, and it actually improves performance as a happy side effect (25% off `InstancerTest.testContextSetPerfWithVariationsParallelEvaluate` for instance), I don't feel the need right now. I'm not sure where the boost has come from, but I expect part of it is that it replaces the very [inefficient way](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/BranchCreator.cpp#L798-L801) we used to find the branch point before.